VEP 168: Masquerade Port Ranges (#167)
kubevirt/enhancements: veps/sig-network/168-masquerade-port-ranges/vep.md

## Goals

- Support configuring entire port ranges to be forwarded to VM guests instead of single specific ones

## Definition of Users

- Any user who wants to expose services or applications running inside a VM guest

## User Stories

- As a Virtual Machine user, I want to specify multiple contiguous ports to forward without having to specify every single one

## Repos

- [KubeVirt](https://github.com/kubevirt/kubevirt/)

## Design

A new `PortRange` type will be added to the KubeVirt API:

```go
// PortRange represents a range of ports to expose from the virtual machine.
// Default protocol TCP.
// The start and end fields are mandatory
type PortRange struct {
	// Protocol for ports. Must be UDP or TCP.
	// Defaults to "TCP".
	// +optional
	Protocol string `json:"protocol,omitempty"`
	// First port of the range to expose for the virtual machine.
	// This must be a valid port number, 0 < x < 65536.
	Start int32 `json:"start"`
  // Last port of the range to expose for the virtual machine.
	// This must be a valid port number, 0 < x < 65536.
  // Must be greater than or equal to start field.
	End int32 `json:"end"`
}
```

> The `PortRanges` field will be added to the existing `Interface` type and made mutually exclusive with the already present `Ports` field:

Reviewer: Simplicity of usage should be the main criteria here.
IMO it should be very simple, as the `Port` version can be easily converted to a `PortRanges` one internally, and optimizing the rules is up to the implementation.
You could phase it, e.g. in Alpha to do it without any optimization and in Beta to change it. But it seems too simple to worth it.

> The `PortRanges` field will be added to the existing `Interface` type and made mutually exclusive with the already present `Ports` field:

Reviewer 2: it looks great now, thanks!

> The `PortRanges` field will be added to the existing `Interface` type and made mutually exclusive with the already present `Ports` field:

Author: Just to clarify, are you approving of the proposed design exposing separate Port and PortRanges fields in the VM/I Spec or do you prefer the "extension" of the already present Port field, as suggested in the Alternatives section?

```go
type Interface struct {
  ...
  // List of ports to be forwarded to the virtual machine.
  // Mutually exclusive with the portRanges field.
	Ports []Port `json:"ports,omitempty"`
  // List of ranges of ports to be forwarded to the virtual machine.
  // Mutually exclusive with the ports field.
	PortRanges []PortRanges `json:"portRanges,omitempty"`
  ...
}
```

## API Examples

```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
spec:
  template:
    spec:
      domain:
        devices:
          interfaces:
            - name: red
              masquerade: {}
              portRanges: # port ranges to forward
                - start: 1000
                  end: 8000
                  protocol: TCP
                - start: 1000
                  end: 8000
                  protocol: UDP
      networks:
      - name: red
        pod: {}
```

> ## Alternatives

Reviewer: I do not see the need to have yet another field, especially in its negative direction.
However, you could just mention it as a considered option and that is it.

An alternative to adding a new field to the `Interface` type, we could instead extend the already present `Ports` field by modifying the `Port` type

### API Design

> A new `PortRange` type will be added to the KubeVirt API:

Reviewer 2: It looks good now, thanks!

      networks:
      - name: red
        pod: {}
```

## Scalability

The proposed modification poses no scalability problems: it can actually make forwarding faster as multiple contiguous ports can be compacted in a single atomic NATting rule instead of multiple ones.

## Update/Rollback Compatibility

**Upgrade:**

- New API field `portRange` is additive and optional
- Existing VMs are unaffected

## Functional Testing Approach

### Unit Tests

- API validation for PortRanges
  - Valid port numbers
  - No overlapping ranges
  - Mutual exclusivity with the Ports field
- Checks for the updated masquerade rules creation logic

### Functional Tests

- PortRanges validation errors are properly reported

### Integration Tests

- Invalid or not correct fields in the templates are rejected with appropriate errors.

## Implementation History

The modification could be carried out in the following steps:
- Schema modifications
- Addition of exhaustive tests of the Schema Validators and the Masquerade Rules
- Modifications to the Schema to Masquerade Rules converters
- Adaptation and extensions of the Schema Validators
- Updates to the Official Documentation

## Graduation Requirements

> Due to the small size of the modification, the entire feature could be developed in either one or two stages (Beta and GA). The backward compatibility of the feature can let us avoid the usage of Feature Gates in the Beta phase and we could leave the Documentation updates to the GA phase.

Reviewer 2: Please consider splitting this into Alpha/Beta/GA sub sections explaining what will be implented/tested in each phase, and which version it will be in (In the alpha section you can mention why that should be skipped)

Reviewer 2: the section looks great now, thanks!

# VEP 168: Masquerade Port Ranges

## VEP Status Metadata

### Target releases

- This VEP targets alpha for version: v1.9
- This VEP targets beta for version: v1.10

> - This VEP targets GA for version: v1.11

Reviewer 3: Per the VEP template guidance, only the upcoming release should be targeted in this section. Please leave the Beta and GA versions blank - we can fill those in during their respective planning phases.
```suggestion
- This VEP targets beta for version:
- This VEP targets GA for version: 
```

## Release Signoff Checklist

Items marked with (R) are required _prior to targeting to a milestone / release_.

- [x] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] (not the initial VEP PR)
- [X] (R) Alpha target version is explicitly mentioned and approved
- [X] (R) Beta target version is explicitly mentioned and approved

> - [X] (R) GA target version is explicitly mentioned and approved

Reviewer 3: ```suggestion
- [] (R) Beta target version is explicitly mentioned and approved
- [] (R) GA target version is explicitly mentioned and approved
```

> ## Overview

Reviewer 3: The overview describes the mechanism but doesn't mention the problem it solves. Consider adding a sentence about the current limitation and its implications on users.

This document describes a mechanism for interfaces of type `masquerade` to forward contiguous port ranges to the VM guests.

> ## Motivation

Reviewer 3: The motivation describes the difficulty of specifying individual ports but doesn't explain why the existing default (forward all traffic when no ports are specified) isn't sufficient. If the user doesn't know which ports they need and wants broad forwarding, the catch-all already provides that.

The PR description and earlier comments paint a clearer picture - IIUC, you need broad-but-not-exhaustive forwarding (everything except certain ports), and since the current API only supports inclusion-mode, you'd have to list every individual port except the excluded ones - which is impractical without ranges.

Could you please adjust the motivation to capture this? It would make the "why" much clearer for future readers.

Currently, when using an interface of type `masquerade` all traffic gets forwarded by default to the VM guest running inside it. The only way to achieve a different behaviour is to configure specific ports in the VM/VMI template (`spec.domain.devices.interfaces.ports` field that gets evaluated when using the masquerade core network binding mentioned before). However, when using KubeVirt to host general purpose desktop VMs, you may not know a priori which ports to forward while still wanting to keep the flexibility of a large set of forwarded ports already available to use. This way we avoid the need to update the VM template and the required guest restart, with all the associated problems (e.g. losing the entire state when using ephemeral instances)

> ## Goals

Reviewer 3: The goal describes the API mechanism ("support configuring port ranges") rather than the outcome. Consider reframing it around the user problem - enabling users to define a broad but bounded set of forwarded ports efficiently when the catch-all is too permissive but individual port listing is impractical.

Please consider adding a goal around reducing the number of nftables rules generated compared to listing equivalent individual ports.

- Support configuring entire port ranges to be forwarded to VM guests instead of single specific ones

> ## Non Goals

Reviewer 3: Consider adding non-goals for:
- Forwarding ports to virt-launcher itself (rather than the guest VM). Capturing it here makes that scoping decision visible to future readers.

-  Istio-aware range handling, as the masquerade code has significant Istio-specific logic.

- Allowing combined use of `Ports` and `PortRanges` in the initial implementation
- Supporting `PortRanges` on non-masquerade bindings
- Supporting `PortRanges` on secondary multus interfaces

## Definition of Users

> - Any user who wants to expose services or applications running inside a VM guest

Reviewer 3: This describes every masquerade user today, including those already using individual `Ports`. Consider narrowing it to the users who actually need this feature - e.g., users running VMs where the set of required ports is not known in advance and the catch-all forwarding is too permissive, making individual port listing impractical.

## User Stories

> - As a Virtual Machine user, I want to specify multiple contiguous ports to forward without having to specify every single one

Reviewer 3: This story describes the API mechanism ("specify multiple contiguous ports without specifying every single one") rather than the user's problem. A script generating 1000 individual `Port` entries would equally satisfy this story.

Please consider the following user story:
- As a user running a general-purpose VM, I want to forward a broad set of ports to the guest while keeping certain ports on the pod, without having to list every allowed port individually which would be impractical to maintain and require a VM spec update and guest restart for each change.

Also, please consider a second story for the operational angle:
- As a cluster administrator, I want port forwarding to scale efficiently so that forwarding thousands of ports doesn't generate thousands of individual nftables rules.

# VEP 168: Masquerade Port Ranges

## VEP Status Metadata

### Target releases

- This VEP targets alpha for version: v1.9
- This VEP targets beta for version:
- This VEP targets GA for version: 

## Release Signoff Checklist

Items marked with (R) are required _prior to targeting to a milestone / release_.

- [x] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] (not the initial VEP PR)
- [X] (R) Alpha target version is explicitly mentioned and approved
- [] (R) Beta target version is explicitly mentioned and approved
- [] (R) GA target version is explicitly mentioned and approved

## Overview

> This document tries to make forwarding more efficient when a user wants to forward contiguous port ranges to the VM guests using `masquerade` interfaces. Currently, the only possibility is to specify an attribute in the VM spec for each single port which can be impractical and hard to maintain from a user perspective. Additionally, this situation also translates into numerous separate nftables rules, leaving efficiency on the table. 

Reviewer 3: nit:
```suggestion
This document tries to make forwarding more efficient when a user wants to forward contiguous port ranges to the VM guests using `masquerade` interfaces, but not all of them. Currently, the only possibility is to specify an attribute in the VM spec for each single port which can be impractical and hard to maintain from a user perspective. Additionally, this situation also translates into numerous separate nftables rules, leaving efficiency on the table. 
```

- If two or more ranges are specified, they must not overlap **unless they use different protocols** (e.g. a TCP range can overlap with a UDP range, but two TCP ranges must not overlap).
- All port numbers must be valid (0 < x < 65536).
- Additional checks may be added for security and consistency as needed.

#### Note on mutual exclusivity

This decision is deliberate: mutual exclusivity between `Ports` and `PortRanges` will be required during Alpha (v1.9) to keep implementation and validation simple and predictable. The team will revisit this constraint during Beta based on Alpha feedback and may allow combined usage (for example: single ports together with ranges) in a later release.

### NFTables rules conversion

The existing logic for converting ports specified in the VM/I specs to nftables rules (currently in `pkg/network/setup/netpod/masquerade/masquerade.go`) will be updated to handle port ranges. For each specified range, a single nftables rule will be created to cover the interval, optimizing the number of rules compared to specifying many single ports.

### Feature Gate

The feature will be guarded by a Feature Gate during Alpha and Beta.

> Proposed Feature Gate name: `MasqueradePortRanges`.

Reviewer 3: Since we may want to use it with other bindings (passt), can this feature gate have a more general name? (`PortRangesSpec` for example)

Reviewer: +1 , but we can change this little one later, not need to block on it.

## API Examples

The following example demonstrates a configuration using both fields. Note that in the Alpha phase, validation will enforce mutual exclusivity, preventing this specific combination. This example also illustrates that ranges of different protocols (TCP and UDP) are allowed to overlap.

```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
spec:
  template:
    spec:
      domain:
        devices:
          interfaces:
            - name: red
              masquerade: {}

>               ports:

Reviewer 4: @g-nitti @orelmisan Hi, it seems like this example contradicts with the previous statement:
```
- `PortRanges` and `Ports` cannot be used together in the same interface (Note: This mutual exclusivity constraint will be enforced for the Alpha release (v1.9) to simplify implementation; it may be relaxed in Beta or later).
```
`Ports` and `PortRanges` cannot live together.
Am I missing something?

Reviewer 3: @enp0s3 please see https://github.com/kubevirt/enhancements/pull/167#discussion_r3149801911 and https://github.com/kubevirt/enhancements/pull/306.

                - name: ssh
                  port: 22

>                   protocol: TCP

Reviewer 3: I would suggest dropping this section, as the VEP declares this combination invalid.

Reviewer: I suggest to fix it as a follow up, so the VEP can be approved before the VEP freeze.

      - start: 1000
        end: 2000
        protocol: TCP
```

Mutual Exclusivity with Inclusion:

Regardless of chosen representation, the Exclusion model is mutually exclusive with the Inclusion model: you must choose either to explicitly list allowed ports (`ports` / `portRanges`) or to start from an open/default policy and list exclusions (`excludedPorts` / `excludedPortRanges`), but not both on the same interface. The two models represent opposite configuration paradigms (allow-list vs. deny-list) and cannot be combined for a single interface.

## Scalability

The proposed modification poses no scalability problems: it can actually make forwarding faster as multiple contiguous ports can be compacted in a single atomic NATting rule instead of multiple ones.

## Update/Rollback Compatibility

**Upgrade:**

- New API field `portRanges` is additive and optional
- Existing VMs are unaffected

## Functional Testing Approach

### Unit Tests

- API validation for PortRanges
  - Valid port numbers
  - Overlap checks (disallowed for same protocol, allowed for different)
  - Mutual exclusivity with the Ports field
- Checks for the updated masquerade rules creation logic

> ### Functional Tests

Reviewer 3: Do you intend to add an end2end test?

Reviewer: Alpha stage does not require it, but it seems natural and basic to cover it as part of the implementation.
As it is Alpha, I would not block on this now, but it will help to see them in the implementation PR.

# VEP 168: Masquerade Port Ranges

## VEP Status Metadata

### Target releases

- This VEP targets alpha for version: v1.9
- This VEP targets beta for version: v1.10
- This VEP targets GA for version: v1.11

## Release Signoff Checklist

Items marked with (R) are required _prior to targeting to a milestone / release_.

- [x] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] (not the initial VEP PR)
- [ ] (R) Alpha target version is explicitly mentioned and approved
- [ ] (R) Beta target version is explicitly mentioned and approved

> - [ ] (R) GA target version is explicitly mentioned and approved

Reviewer 2: You can probably `[X]` all of these as they are now clearly stated in `Target releases`.
